[pytx] Add Tech Against Terrorism Hash API

Bruce-Pouncey-TAT opened the pull request:
Summary
---------
Hi

Bruce here from Tech Against Terrorism 👋 

This PR contains Step 1 from [this issue](https://github.com/facebook/ThreatExchange/issues/1610) integrating the Tech Against Terrorism Hash List API into `python-threatexchange`

I used StopNSCII & NCMEC inside `/client/exchanges/` as initial guidance. 

**Caveat**
Our API returns a pre-signed URL to a JSON file for our hash list, instead of paginated results. I'm wondering how we want to handle this moving forward into part 2 (Once we're happy with part 1) ? 

Test Plan
---------
**Manual**
For manual testing you can use python shell to simply instantiate the `TATHashListAPI` with your TCAP username and password.

```
tat_api = TATHashListAPI(username=USER_NAME, password=PASSWORD)
response = tat_api.get_hash_list()
```

**Automated**
I have written unit tests `techagainstterrorism/tests` so `pytest` will do it.

Dcallies commented:
Hey Bruce! Great to be working with you! I will help you get this landed and review it.

As for the details of the API, I did read your dos and though you picked a bit of a challenging implementation, but I think we can make it work! I'll give you some tips on the PR itself.

Bruce-Pouncey-TAT commented:
Hey David, likewise! 
I will get the pipeline failures resolved in the coming days,  I'm also waiting on the CLA confirmation to come through on my side.

Dcallies reviewed (changes requested):
Overall looks pretty good! 

Thank you for including your manual test steps in the PR summary!

The one blocking thing I have is that I think the interface will become easier if you standardize on throwing exceptions instead of catching them and return a new object. 

Let me know if you are short on time or energy, I'm willing to do the labor to bring this in the rest of the way if you are short on either of those.

Bruce-Pouncey-TAT commented on python-threatexchange/threatexchange/exchanges/clients/techagainstterrorism/api.py:
```diff
@@ -0,0 +1,169 @@
+
+    This API delivers a JSON file containing a comprehensive list of all hashed terrorist content within the TAT system.
+
+    The list is refreshed daily.
```
I have added links to our public docs on how our OSINT team operate as well as docs linking to out Hash List API.

Dcallies replied:
Looks great!

Bruce-Pouncey-TAT commented on python-threatexchange/threatexchange/exchanges/clients/techagainstterrorism/api.py:
```diff
@@ -0,0 +1,175 @@
+
+@dataclass
+class TATUser:
+  user: Any
```
Our Actual User object can be quite large (possibly hundreds of permissions) and will vary in size based on who is using the interface. I can arrow this down to `Dict[str, Any]`

Bruce-Pouncey-TAT commented:
> Overall looks pretty good!
> 
> Thank you for including your manual test steps in the PR summary!
> 
> The one blocking thing I have is that I think the interface will become easier if you standardize on throwing exceptions instead of catching them and return a new object.
> 
> Let me know if you are short on time or energy, I'm willing to do the labor to bring this in the rest of the way if you are short on either of those.

Thanks!
I have resolved the comments with my best understanding of what you have requested. Also I have the time and energy to continue with anymore changes you suggest.

I have moved the `TimeoutHTTPAdapter` into a common utils file and updated where it has been used.

**Idea** 🤔 
I have noticed across all clients we're using `_get_session` within a content manager
Maybe we could lift this up into a parent class ? `PyTxClientAPIBase` for the lack of a better definition or maybe introduce an Abstract Base Class. Just a thought 😄

Dcallies commented on python-threatexchange/threatexchange/exchanges/clients/utils/common.py:
```diff
@@ -0,0 +1,17 @@
+from requests.adapters import HTTPAdapter
```
nit: I git an internal nagger that forces me to add 

`# Copyright (c) Meta Platforms, Inc. and affiliates.` to the tops of all files, though I can do it as a followup

Dcallies commented on python-threatexchange/threatexchange/exchanges/clients/techagainstterrorism/tests/test_api.py:
```diff
@@ -0,0 +1,94 @@
+import typing as t
```
nit: I git an internal nagger that forces me to add 

`# Copyright (c) Meta Platforms, Inc. and affiliates.` to the tops of all files, though I can do it as a followup

Dcallies commented on python-threatexchange/threatexchange/exchanges/clients/techagainstterrorism/api.py:
```diff
@@ -0,0 +1,152 @@
+
+        except Exception as exception:
+            logging.error("Failed to get hash list: %s", exception)
+            raise Exception("Failed to fetch Hash List: %s", exception)
```
blocking (though I'm fine fixing it on a followup): Instead of wrapping this exception, rethrow the one on stack by just doing 

```
raise
```

Ref: https://docs.python.org/3/tutorial/errors.html#enriching-exceptions-with-notes

Bruce-Pouncey-TAT replied:
I was in the code so I made the change. Thanks for the reference!

Dcallies reviewed (changes requested):
> Idea 🤔
> I have noticed across all clients we're using _get_session within a content manager
Maybe we could lift this up into a parent class ? PyTxClientAPIBase for the lack of a better definition or maybe introduce an Abstract Base Class for the places where client are using _get _post. Just a thought 😄

It's an interesting idea! Let's first get your SignalExchangeAPI implementation working, and we can come back to it.

You only have ONE thing left to fix - the lint error on running black!

Dcallies reviewed (approved):
Looks good!
